exporters/metrics/util/src/main/java/io/opencensus/exporter/metrics/util/QueueMetricProducer.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2019, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.exporter.metrics.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Scope;
+import io.opencensus.metrics.export.Metric;
+import io.opencensus.metrics.export.MetricProducer;
+import io.opencensus.trace.AttributeValue;
+import io.opencensus.trace.Sampler;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import io.opencensus.trace.samplers.Samplers;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Queue;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Wrapper of {@link MetricProducer} which allows metrics to be pushed and buffered.
+ *
+ * @since 0.20
+ */
+public final class QueueMetricProducer extends MetricProducer {
+
+  private static final Object monitor = new Object();
+  private static final Tracer tracer = Tracing.getTracer();
+  private static final Sampler probabilitySampler = Samplers.probabilitySampler(0.0001);
+
+  private static final String DEFAULT_SPAN_NAME = "PushMetrics";
+  private static final int DEFAULT_BUFFER_SIZE = 32;
+  private static final String ATTRIBUTE_KEY_DROPPED_METRICS = "DroppedMetrics";
+
+  @GuardedBy("monitor")
+  private final Queue<Metric> bufferedMetrics = new LinkedList<>();
```

Comment:
Maybe consider to use https://google.github.io/guava/releases/23.0/api/docs/com/google/common/collect/EvictingQueue.html

Reply:
SG, updated.